zipkin-ui/js/component_ui/traceToMustache.js

```diff
@@ -103,96 +33,116 @@ function incrementServiceNameCount(span, results) {
-  // make sense to align this with service depth as that's what the dependency graph shows. Either
-  // that or remote depth.. eg what's the longest chain of remote spans (taking care to ignore
-  // redundant instrumentation.
-  const depth = tree.maxDepth();
-
-  const spans = _(getRootSpans(trace)).flatMap(
-    (rootSpan) => childrenToList(createSpanTreeEntry(rootSpan, trace))).map((span) => {
-      const spanStartTs = span.timestamp || traceTimestamp;
-      const spanDepth = spanDepths[span.id] || 1;
-      const spanDuration = span.duration || 0;
-      const children = (groupByParentId[span.id] || []).map((s) => s.id);
-
-      const res = {
-        spanId: span.id,
-        depth: (spanDepth + 1) * 5,
-        depthClass: (spanDepth - 1) % 6,
-        annotations: span.annotations.map((a) => ({
-          ...a,
-          left: spanDuration ? (a.timestamp - spanStartTs) / spanDuration * 100 : 0,
-          relativeTime: mkDurationStr(a.timestamp - traceTimestamp),
-          width: 8
-        })),
-        tags: span.tags,
-        errorType: getErrorType(span)
-      };
-
-      // Optionally add fields instead of defaulting to empty string
-      if (span.name) res.spanName = span.name;
-      if (spanDuration) {
-        const width = traceDuration ? spanDuration / traceDuration * 100 : 0;
-        res.width = width < 0.1 ? 0.1 : width;
-        res.left = parseFloat(spanStartTs - traceTimestamp) / parseFloat(traceDuration) * 100;
-        res.duration = spanDuration; // used in zoom
-        res.durationStr = mkDurationStr(spanDuration); // bubble over the span in trace view
+export function traceToMustache(root, logsUrl) {
+  const spans = root.traverse();
+  if (spans.length === 0) throw new Error('Trace was empty');
+  const traceTimestamp = spans[0].timestamp;
+  if (!traceTimestamp) throw new Error('Trace is missing a timestamp');
+
+  // currently we need to pass one traversal in order to get the duration
+  const traceDuration = getTraceDuration(spans);
+
+  const modelview = {
+    traceId: spans[0].traceId,
+    depth: 0,
+    spanCount: spans.length,
+    serviceNameAndSpanCounts: getServiceNameAndSpanCounts(spans),
+    spans: []
+  };
+
+  const queue = [];
+
+  // since the input data could be headless, we first push onto the queue the root-most spans
+  if (typeof(root.span) === 'undefined') { // synthetic root
+    root.children.forEach(child => queue.push(child));
+  } else {
+    queue.push(root);
+  }
+
+  while (queue.length > 0) {
+    let current = queue.shift();
+
+    let span = SPAN_V1.convert(current.span);
+
+    // This is more than a normal tree traversal, as we are merging any server spans that share the
+    // same ID. When that's the case, we pull up any of their children as if they are our own.
+    const childIds = [];
+    current.children.forEach(child => {
+      if (child.span.id === span.id) {
+        span = SPAN_V1.merge(span, SPAN_V1.convert(child.span));
```

Comment:
ps in a later change we can even pull up servers to fix the single host span issue